client/helpers_test.go

```diff
@@ -147,3 +148,431 @@ func TestApplyChangelistMulti(t *testing.T) {
+	assert.Equal(t, "level2", role2.Paths[0])
+
+	// delete one delegation, ensure key remains
+	ch = changelist.NewTufChange(
+		changelist.ActionDelete,
+		"targets/level1",
+		changelist.TypeTargetsDelegation,
+		"",
+		nil,
+	)
+
+	err = applyTargetsChange(repo, ch)
+	assert.NoError(t, err)
+
+	assert.Len(t, tgts.Signed.Delegations.Roles, 1)
+	assert.Len(t, tgts.Signed.Delegations.Keys, 1)
+
+	// delete other delegation, ensure key cleaned up
+	ch = changelist.NewTufChange(
+		changelist.ActionDelete,
+		"targets/level2",
+		changelist.TypeTargetsDelegation,
+		"",
+		nil,
+	)
+
+	err = applyTargetsChange(repo, ch)
+	assert.NoError(t, err)
+
+	assert.Len(t, tgts.Signed.Delegations.Roles, 0)
+	assert.Len(t, tgts.Signed.Delegations.Keys, 0)
+}
+
+func TestApplyTargetsDelegationCreateEdit(t *testing.T) {
+	_, repo, cs := testutils.EmptyRepo()
+
+	newKey, err := cs.Create("targets/level1", data.ED25519Key)
+	assert.NoError(t, err)
+
+	// create delegation
+	kl := data.KeyList{newKey}
+	td := &changelist.TufDelegation{
+		NewThreshold: 1,
+		AddKeys:      kl,
+		AddPaths:     []string{"level1"},
+	}
+
+	tdJSON, err := json.Marshal(td)
+	assert.NoError(t, err)
+
+	ch := changelist.NewTufChange(
+		changelist.ActionCreate,
+		"targets/level1",
+		changelist.TypeTargetsDelegation,
+		"",
+		tdJSON,
+	)
+
+	err = applyTargetsChange(repo, ch)
+	assert.NoError(t, err)
+
+	// edit delegation
+	newKey2, err := cs.Create("targets/level1", data.ED25519Key)
+	assert.NoError(t, err)
+
+	kl = data.KeyList{newKey2}
+	td = &changelist.TufDelegation{
+		NewThreshold: 1,
+		AddKeys:      kl,
+		RemoveKeys:   []string{newKey.ID()},
+	}
+
+	tdJSON, err = json.Marshal(td)
+	assert.NoError(t, err)
+
+	ch = changelist.NewTufChange(
+		changelist.ActionUpdate,
+		"targets/level1",
+		changelist.TypeTargetsDelegation,
+		"",
+		tdJSON,
+	)
+
+	err = applyTargetsChange(repo, ch)
+	assert.NoError(t, err)
+
+	tgts := repo.Targets[data.CanonicalTargetsRole]
+	assert.Len(t, tgts.Signed.Delegations.Roles, 1)
+	assert.Len(t, tgts.Signed.Delegations.Keys, 1)
+
+	_, ok := tgts.Signed.Delegations.Keys[newKey2.ID()]
+	assert.True(t, ok)
+
+	role := tgts.Signed.Delegations.Roles[0]
+	assert.Len(t, role.KeyIDs, 1)
+	assert.Equal(t, newKey2.ID(), role.KeyIDs[0])
+	assert.Equal(t, "targets/level1", role.Name)
+	assert.Equal(t, "level1", role.Paths[0])
+}
+
+func TestApplyTargetsDelegationInvalidRole(t *testing.T) {
+	_, repo, cs := testutils.EmptyRepo()
+
+	newKey, err := cs.Create("targets/level1", data.ED25519Key)
+	assert.NoError(t, err)
+
+	// create delegation
+	kl := data.KeyList{newKey}
+	td := &changelist.TufDelegation{
+		NewThreshold: 1,
+		AddKeys:      kl,
+		AddPaths:     []string{"level1"},
+	}
+
+	tdJSON, err := json.Marshal(td)
+	assert.NoError(t, err)
+
+	ch := changelist.NewTufChange(
+		changelist.ActionCreate,
+		"bad role",
+		changelist.TypeTargetsDelegation,
+		"",
+		tdJSON,
+	)
+
+	err = applyTargetsChange(repo, ch)
+	assert.Error(t, err)
```

Comment:
Would it make sense to assert the `ErrInvalidRole` type here, so we don't conflate with some other error?